mesh/ncnurbs.cpp

```diff
@@ -0,0 +1,3344 @@
+{
+   perm.SetSize(n1 * n2);
+
+   // The ordering of entities in the face is based on the vertices.
+
+   Array<int> faceEdges, ori, evert, e2vert, vert;
+   patchTopo->GetFaceEdges(sf, faceEdges, ori);
+   patchTopo->GetFaceVertices(sf, vert);
+   patchTopo->GetEdgeVertices(faceEdges[e1], evert);
+   MFEM_ASSERT(evert[0] == v0 || evert[1] == v0, "");
+
+   bool d[2];
+   d[0] = (evert[0] == v0);
+
+   const int v10 = d[0] ? evert[1] : evert[0];
+
+   // The face has {fn1,fn2} interior entities, with ordering based on `vert`.
+   // Now we find these sizes by first finding the edge with vertices [v0, v10].
+   int e0 = -1;
+   for (int i=0; i<4; ++i)
+   {
+      patchTopo->GetEdgeVertices(faceEdges[i], evert);
+      if ((evert[0] == v0 && evert[1] == v10) ||
+          (evert[1] == v0 && evert[0] == v10)) { e0 = i; }
+   }
+
+   MFEM_ASSERT(e0 >= 0, "");
+
+   const bool tr = e0 % 2 == 1; // True means (fn1,fn2) == (n2,n1)
+
+   patchTopo->GetEdgeVertices(faceEdges[e2], evert);
+   MFEM_ASSERT(evert[0] == v10 || evert[1] == v10, "");
+   d[1] = (evert[0] == v10);
+
+   const int v11 = d[1] ? evert[1] : evert[0];
+
+   int v01 = -1;
+   for (int i=0; i<4; ++i)
+   {
+      if (vert[i] != v0 && vert[i] != v10 && vert[i] != v11) { v01 = vert[i]; }
+   }
+
+   MFEM_ASSERT(v01 >= 0 && v01 == vert.Sum() - v0 - v10 - v11, "");
+
+   // Translate indices [v0, v10, v11, v01] to pairs of indices in {0,1}.
+   constexpr char ipair[4][2] = {{0, 0}, {1, 0}, {1, 1}, {0, 1}};
+   int f00[2];
+
+   int allv[4] = {v0, v10, v11, v01};
+   int locv[4];
+   for (int i=0; i<4; ++i)
+   {
+      locv[i] = -1;
+      for (int j=0; j<4; ++j)
+      {
+         if (vert[j] == allv[i])
+         {
+            locv[i] = j;
+         }
+      }
+
+      MFEM_ASSERT(locv[i] >= 0, "");
+   }
+
+   for (int i=0; i<2; ++i) { f00[i] = ipair[locv[0]][i]; }
+
+   const int i0 = f00[0];
+   const int j0 = f00[1];
+
+   for (int i=0; i<n1; ++i)
+      for (int j=0; j<n2; ++j)
+      {
+         // Entity perm[i] of the face should be entity i in the master face
+         // ordering. The master face ordering varies faster in the direction
+         // from v0 to v10, and slower in the direction from v10 to v11, or
+         // equivalently, from v0 to v01.
+         if (tr)
+         {
+            const int fi = i0 == 0 ? j : n2 - 1 - j;
+            const int fj = j0 == 0 ? i : n1 - 1 - i;
+            const int p = fi + (fj * n2); // Index in the slave face ordering
+            const int m = i + (j * n1); // Index in the master face ordering
+            perm[m] = p;
+         }
+         else
+         {
+            const int fi = i0 == 0 ? i : n1 - 1 - i;
+            const int fj = j0 == 0 ? j : n2 - 1 - j;
+            const int p = fi + (fj * n1); // Index in the slave face ordering
+            const int m = i + (j * n1); // Index in the master face ordering
+            perm[m] = p;
+         }
+      }
+}
+
+bool ConsistentlySetEntry(int v, int &e)
```

Comment:
This function and others below could use some documentation.

Reply:
Done in 9d3047b3a77ce40300aeefea3c93be11199ec298